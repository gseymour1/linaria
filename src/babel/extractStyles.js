/* @flow */

// import sheet from '../sheet';

export default function extractStyles() {
  // @TODO
}
