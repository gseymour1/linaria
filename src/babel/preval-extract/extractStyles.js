/* @flow */

import fs from 'fs';
import path from 'path';

import type { BabelTypes, NodePath } from '../types';

import { getCachedModule } from '../lib/moduleSystem';

const preamble =
  '/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */';

function parseCurrentFilename(filename: string, outDir: string = '') {
  const dirname = path.isAbsolute(filename)
    ? path.dirname(filename)
    : path.join(process.cwd(), outDir);
  const basename = /(.+)\..+$/.exec(path.basename(filename))[1];
  return path.join(dirname, `${basename}.css`);
}

function parseFilename(
  filename: string,
  currentFilename: string,
  outDir: string = ''
) {
  const dirname = path.isAbsolute(outDir)
    ? outDir
    : path.join(process.cwd(), outDir);
  const basename = /(.+)\..+$/.exec(path.basename(currentFilename))[1];
  return path.join(dirname, filename.replace('[name]', basename));
}

let stylesCache = {};

function hasCachedStyles(filename: string, styles: string) {
  return stylesCache[filename] && stylesCache[filename] === styles;
}

function createCssFromCache() {
  return Object.keys(stylesCache).reduce(
    (acc, file) => `${acc}\n${stylesCache[file]}`,
    preamble
  );
}

function addRequireForCss(
  types: BabelTypes,
  program: NodePath<*>,
  filename: string
) {
  program.node.body.unshift(
    types.expressionStatement(
      types.callExpression(types.identifier('require'), [
        types.stringLiteral(filename),
      ])
    )
  );
}

export default function extractStyles(
  types: BabelTypes,
  program: NodePath<*>,
  currentFilename: string,
  options: {
    single?: boolean,
    filename?: string,
    outDir?: string,
    cache?: boolean,
    extract?: boolean,
  }
) {
  const { single, filename, cache, extract } = {
    cache: true,
    extract: true,
    single: false,
    ...options,
    filename: options.filename
      ? parseFilename(options.filename, currentFilename, options.outDir)
      : parseCurrentFilename(currentFilename, options.outDir),
  };

  if (!extract) {
    return;
  }

  const sheet = getCachedModule(require.resolve('../../sheet.js'));
  const data = sheet ? sheet.exports.default.dump() : '';

  if (!data.length) {
    return;
  }

  if (cache) {
    if (single) {
      if (hasCachedStyles(currentFilename, data)) {
        return;
      }
      stylesCache[currentFilename] = data;
    } else {
      if (hasCachedStyles(filename, data)) {
        addRequireForCss(types, program, filename);
        return;
      }
      stylesCache[filename] = data;
    }
  }

  if (single) {
    fs.writeFileSync(filename, createCssFromCache());
  } else {
    fs.writeFileSync(filename, `${preamble}\n${data}`);
    addRequireForCss(types, program, filename);
  }
}

export function clearCache() {
  stylesCache = {};
}
