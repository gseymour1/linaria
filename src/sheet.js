/* @flow */

import stylis from 'stylis';

function sheet() {
  let ruleCache: ?Array<CSSRule>;
  let cssText: ?string;
  let stylesCache: ?Array<{ selector: string, css: string }>;

  if (typeof window === 'undefined' || typeof window.document === 'undefined') {
    return {
      insert(selector: string, css: string) {
        const text = stylis(selector, css);
        stylesCache = stylesCache || [];
        stylesCache.push({ selector, css });
        cssText = cssText ? cssText + text : text;
      },
      rules() {
        throw new Error('Not implemented');
      },
      styles() {
        return stylesCache || [];
      },
      dump() {
        const result = cssText || '';
        cssText = null;
        stylesCache = null;
        return result;
      },
    };
  }

  const style = document.createElement('style');
  const node = document.createTextNode('');

  style.appendChild(node);
  style.setAttribute('type', 'text/css');

  if (document.head != null) {
    document.head.appendChild(style);
  } else {
    throw new Error('Unable to insert stylesheet');
  }

  return {
    insert(selector: string, css: string) {
      const text = stylis(selector, css);
      node.appendData(text);
      // invalidate the cache since stylesheets have changed
      ruleCache = null;
    },
    rules() {
      if (ruleCache != null) {
        return { list: ruleCache, cache: true };
      }

      ruleCache = [].concat(
        /* $FlowFixMe */
        ...Array.from(document.styleSheets).map(s => Array.from(s.cssRules))
      );

      return { list: ruleCache, cache: false };
    },
  };
}

export default sheet();
