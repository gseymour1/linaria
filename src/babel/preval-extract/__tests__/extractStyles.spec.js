import path from 'path';
import fs from 'fs';
import { types } from 'babel-core';
import extractStyles, { clearCache } from '../extractStyles';
import sheet from '../../../sheet';

jest.mock('../../../sheet');
jest.mock('fs');

describe('extractStyles module from preval-extract babel plugin', () => {
  it('should not do anything if there is no styles', () => {
    sheet.dump.mockImplementationOnce(() => '');
    const writeFileSync = jest.fn();

    extractStyles(types, null, '', {}, { writeFileSync });

    sheet.dump.mockImplementationOnce(() => '');
    extractStyles(types, null, '', {});

    expect(writeFileSync).not.toHaveBeenCalled();
    expect(fs.writeFileSync).not.toHaveBeenCalled();
  });

  it('should skip extraction if extract option is false', () => {
    sheet.dump.mockImplementationOnce(() => '...');
    const writeFileSync = jest.fn();

    extractStyles(
      types,
      null,
      'filename.js',
      { extract: false },
      { writeFileSync }
    );

    expect(writeFileSync).not.toHaveBeenCalled();
  });

  describe('with cache disabled', () => {
    beforeEach(clearCache);

    it('should append new styles to a file', () => {
      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');
      const writeFileSync = jest.fn();

      const program = {
        node: {
          body: [],
        },
      };
      extractStyles(
        types,
        program,
        'filename.js',
        { single: true, cache: false },
        { writeFileSync }
      );

      expect(writeFileSync).toHaveBeenCalledTimes(1);
      expect(writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), 'filename.css')
      );
      expect(writeFileSync.mock.calls[0][1]).toMatchSnapshot();
      expect(program.node.body.length).toBe(0);
    });

    it('should write (and overwrite) new styles to a file', () => {
      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');
      const writeFileSync = jest.fn();

      const program = {
        node: {
          body: [],
        },
      };

      extractStyles(
        types,
        program,
        path.join(process.cwd(), 'filename.js'),
        { cache: false },
        { writeFileSync }
      );

      sheet.dump.mockImplementationOnce(() => '.classname{color: #000000}');
      // On each transpilation the body does not contain require call added by extractStyles
      program.node.body = [];

      extractStyles(
        types,
        program,
        path.join(process.cwd(), 'filename.js'),
        { cache: false },
        { writeFileSync }
      );

      expect(writeFileSync).toHaveBeenCalledTimes(2);
      expect(writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), 'filename.css')
      );
      expect(writeFileSync.mock.calls[0][1]).toMatchSnapshot();
      expect(program.node.body.length).toBe(1);
    });
  });

  describe('with cache enabled', () => {
    beforeEach(clearCache);

    it('should write only once if the file has not changed', () => {
      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');
      const writeFileSync = jest.fn();

      const program = {
        node: {
          body: [],
        },
      };

      extractStyles(
        types,
        program,
        'filename.js',
        { filename: 'A.css' },
        { writeFileSync }
      );

      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');

      extractStyles(
        types,
        program,
        'filename.js',
        { filename: 'A.css' },
        { writeFileSync }
      );

      expect(writeFileSync).toHaveBeenCalledTimes(1);
      expect(writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), 'A.css')
      );
      expect(writeFileSync.mock.calls[0][1]).toMatchSnapshot();
      expect(program.node.body.length).toBe(1);
    });

    it('should overwrite if the file has changed', () => {
      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');
      const writeFileSync = jest.fn();

      const program = {
        node: {
          body: [],
        },
      };

      extractStyles(
        types,
        program,
        'filename.js',
        { outDir: process.cwd(), filename: '[name].css' },
        { writeFileSync }
      );

      sheet.dump.mockImplementationOnce(() => '.classname{color: #000000}');
      // On each transpilation the body does not contain require call added by extractStyles
      program.node.body = [];

      extractStyles(
        types,
        program,
        'filename.js',
        { outDir: process.cwd(), filename: '[name].css' },
        { writeFileSync }
      );

      expect(writeFileSync).toHaveBeenCalledTimes(2);
      expect(writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), 'filename.css')
      );
      expect(writeFileSync.mock.calls[0][1]).toMatch('#ffffff');
      expect(writeFileSync.mock.calls[0][1]).toMatchSnapshot();
      expect(writeFileSync.mock.calls[1][0]).toEqual(
        path.join(process.cwd(), 'filename.css')
      );
      expect(writeFileSync.mock.calls[1][1]).toMatch('#000000');
      expect(writeFileSync.mock.calls[1][1]).toMatchSnapshot();
      expect(program.node.body.length).toBe(1);
    });

    it('should append styles only once if the file has not changed', () => {
      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');
      const writeFileSync = jest.fn();

      extractStyles(
        types,
        null,
        'filename.js',
        { single: true },
        { writeFileSync }
      );

      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');

      extractStyles(
        types,
        null,
        'filename.js',
        { single: true },
        { writeFileSync }
      );

      expect(writeFileSync).toHaveBeenCalledTimes(1);
      expect(writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), 'filename.css')
      );
      expect(writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(writeFileSync.mock.calls[0][1]).toMatchSnapshot();
    });

    it('should overwrite css if the file has changed', () => {
      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');
      const writeFileSync = jest.fn();

      extractStyles(
        types,
        null,
        'filename.js',
        { single: true },
        { writeFileSync }
      );

      sheet.dump.mockImplementationOnce(() => '.classname{color: #000000}');

      extractStyles(
        types,
        null,
        'filename.js',
        { single: true },
        { writeFileSync }
      );

      expect(writeFileSync).toHaveBeenCalledTimes(2);
      expect(writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), 'filename.css')
      );
      expect(writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(writeFileSync.mock.calls[0][1]).toMatchSnapshot();
      expect(writeFileSync.mock.calls[1][0]).toEqual(
        path.join(process.cwd(), 'filename.css')
      );
      expect(writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #000000}'
      );
      expect(writeFileSync.mock.calls[1][1]).toMatchSnapshot();
    });

    it('should overwrite css if the new file was added', () => {
      sheet.dump.mockImplementationOnce(() => '.classname{color: #ffffff}');
      const writeFileSync = jest.fn();

      extractStyles(
        types,
        null,
        'filename1.js',
        { single: true },
        { writeFileSync }
      );

      sheet.dump.mockImplementationOnce(() => '.classname{color: #000000}');

      extractStyles(
        types,
        null,
        'filename2.js',
        { single: true },
        { writeFileSync }
      );

      expect(writeFileSync).toHaveBeenCalledTimes(2);
      expect(writeFileSync.mock.calls[0][0]).toEqual(
        path.join(process.cwd(), 'filename1.css')
      );
      expect(writeFileSync.mock.calls[0][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(writeFileSync.mock.calls[0][1]).toMatchSnapshot();
      expect(writeFileSync.mock.calls[1][0]).toEqual(
        path.join(process.cwd(), 'filename2.css')
      );
      expect(writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #ffffff}'
      );
      expect(writeFileSync.mock.calls[1][1]).toMatch(
        '.classname{color: #000000}'
      );
      expect(writeFileSync.mock.calls[1][1]).toMatchSnapshot();
    });
  });
});
