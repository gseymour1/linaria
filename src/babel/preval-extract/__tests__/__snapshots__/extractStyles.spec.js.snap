// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval-extract/extractStyles module with cache disabled with single set to true, should write styles to a file 1`] = `
"/**
 * THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY OR COMMIT IT TO VERSION CONTROL.
 * SOURCE: filename.js
 */

.classname{color: #ffffff}"
`;
